FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1323 @@
+            } else {
+              IREmit->_StoreMem(FPRClass, Op->StoreSize, AddrNode, StackNode);
+            }
+          }
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_STORESTACKTOSTACK: { // stores top of stack in another place in stack.
+        const auto* Op = IROp->C<IROp_StoreStackToStack>();
+
+        auto Offset = Op->StackLocation;
+
+        if (Offset != 0) {
+          const auto& [Valid, Value] = StackData.top();
+          MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+          // Need to store st0 to stack location - basically a copy.
+          if (SlowPath) {
+            StoreStackValueAtOffset_Slow(Offset, LoadStackValueAtTop_Slow());
+          } else {
+            StackData.setTop(Value, Offset);
+          }
+        }
+
+        IREmit->Remove(CodeNode);
+        break;
+      }
+      case OP_POPSTACKDESTROY: {
+        if (SlowPath) {
+          SetX87ValidTag(GetTopWithCache_Slow(), false);
+          UpdateTopForPop_Slow();
+        } else {
+          StackData.pop();
+        }
+
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80STACKXCHANGE: {
+        const auto* Op = IROp->C<IROp_F80StackXchange>();
+        auto Offset = Op->SrcStack;
+
+        const auto& [ValidTop, StackTop] = StackData.top();
+        const auto& [Valid, StackMember] = StackData.top(Offset);
+
+        MigrateToSlowPathIf(ValidTop != StackSlot::VALID || Valid != StackSlot::VALID);
+
+        if (SlowPath) { // slow path
+          auto* ValueTop = LoadStackValueAtTop_Slow();
+          auto* ValueOffset = LoadStackValueAtOffset_Slow(Offset);
+          StoreStackValueAtTop_Slow(ValueOffset, false);
+          StoreStackValueAtOffset_Slow(Offset, ValueTop, false);
+        } else { // fast path
+          auto Tmp = StackTop;
+          StackData.setTop(StackMember);
+          StackData.setTop(Tmp, Offset);
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80STACKCHANGESIGN: {
+        const auto& [Valid, StackMember] = StackData.top();
+
+        // We need a couple of intermediate instructions to change the sign
+        // of a value
+        Ref HelperNode {};
+        if (!ReducedPrecisionMode) {
+          Ref Low = GetConstant(0);
+          Ref High = GetConstant(0b1'000'0000'0000'0000ULL);
+          HelperNode = IREmit->_VCastFromGPR(16, 8, Low);
+          HelperNode = IREmit->_VInsGPR(16, 8, 1, HelperNode, High);
+        }
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        if (SlowPath) { // slow path
+          Ref ResultNode {};
+          Ref Value = LoadStackValueAtTop_Slow();
+          // Negate value
+          if (ReducedPrecisionMode) {
+            ResultNode = IREmit->_VFNeg(8, 8, Value);
+          } else {
+            ResultNode = IREmit->_VXor(16, 1, Value, HelperNode);
+          }
+          StoreStackValueAtTop_Slow(ResultNode, false);
```

Comment:
I think we want to extract a `SetTop(Ref)` helper that does

```
if (SlowPath) {
   StoreStackValueAtTop_Slow(Ref, false);
} else {
   StackData.setTop(StackMemberInfo {
            .SourceDataSize = StackMember.SourceDataSize,
            .SourceDataNode = nullptr,
            .StackDataNode = Ref
   });
}
```

Along with the `LoadStackTop()` helper I proposed below, this will make this if/else go away and deduplicate the logic.